### PR TITLE
MARKET-1672 World Clock component

### PR DESCRIPTION
**Pull request by Fialkaksu**
The World Clock component provides several options for displaying time in different time zones, as well as options for setting the time zone and labels for them.

[MARKET-1672](https://bugs.backendless.com/browse/MARKET-1672)
[UI-Builder](https://develop.backendless.com/app/TestComponentsApp/ui-builder/WorldClock/pages/landing)
[Desktop Preview](https://populardowntown.backendless.app/api/files/ui-builder/containers/WorldClock/index.html?page=landing)

image

### CODE REVIEW

File: components/bl-worldClock-component/src/digital-subtypes/DigitalClockPixel.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      .map(() => `rgb(0, ${Math.floor(Math.random() * 150)}, ${Math.floor(Math.random() * 255)})`);
+    setColors(newColors);
+  }, []);
+
+  const digits = useMemo(() => {
+    const [hour, minute, second] = time;
+    const formattedTime = displaySeconds ? [...hour, ':', ...minute, ':', ...second] : [...hour, ':', ...minute];
+
+    return formattedTime;
+  }, [time]);
```
Comment (serhmel) [inline components/bl-worldClock-component/src/digital-subtypes/DigitalClockPixel.js:48]: Is it not the same?
```suggestion
  const digits = useMemo(() => {
    const [hour, minute, second] = time;

    return displaySeconds ? [hour, ':', minute, ':', second] : [hour, ':', minute];
  }, [time]);
```

File: components/bl-worldClock-component/src/digital-subtypes/DigitalClockPixel.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  return (
+    <div className="clock">
+      { digits.map((value, index) => (
+        <Digit key={ index } value={ value } colors={ colors } />
+      )) }
+
+      { isAmpm && ( <span className="ampm">{ ampm }</span> ) }
+    </div>
```
Comment (serhmel) [inline components/bl-worldClock-component/src/digital-subtypes/DigitalClockPixel.js:57]:
```suggestion
    <div className="clock">
      { digits.map((value, index) => (
        <Digit key={ index } value={ value } colors={ colors }/>
      )) }

      { isAmpm && (
        <span className="ampm">{ ampm }</span>
      ) }
    </div>
```

File: components/bl-worldClock-component/src/views/digital-types/DigitalClockFlip.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    if (ampm !== state.ampm) {
+      setState(prevState => ({ ...prevState, ampm, ampmShuffle: !state.ampmShuffle }));
+    }
+  }, [hour, minute, second, ampm]);
+
+  return (
+    <div className="flip-clock">
+      <FlipUnitContainer unit="hour" digit={ state.hour } shuffle={ state.hourShuffle }/>
+      <Semicolon />
```
Comment (serhmel) [inline components/bl-worldClock-component/src/views/digital-types/DigitalClockFlip.js:87]:
```suggestion
      <Semicolon/>
```

Below the same.

File: components/bl-worldClock-component/src/views/digital-types/DigitalClockPixel.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    <div className="digit">
+      { digits[value].map((active, index) => (
+        <div className="segment" style={{ backgroundColor: active ? colors[index] : 'transparent' }} key={ index }/>
+      )) }
+    </div>
+  );
+}
+
+export function DigitalClockPixel({ time, displaySeconds }) {
+  const { hour, minute, second, isAmpm, ampm } = time;
```
Comment (serhmel) [inline components/bl-worldClock-component/src/views/digital-types/DigitalClockPixel.js:28]: empty line?

File: components/bl-worldClock-component/src/views/digital-types/DigitalClockPixel.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    const newColors = Array(15)
+      .fill()
+      .map(() => `rgb(0, ${ Math.floor(Math.random() * 150) }, ${ Math.floor(Math.random() * 255) })`);
+
+    setColors(newColors);
+  }, []);
+
+  const digits = useMemo(() => {
+    return displaySeconds ? [...hour, ':', ...minute, ':', ...second] : [...hour, ':', ...minute];
+  }, [hour, minute, second]);
```
Comment (serhmel) [inline components/bl-worldClock-component/src/views/digital-types/DigitalClockPixel.js:41]:
```suggestion
  const digits = useMemo(() => (
    displaySeconds ? [...hour, ':', ...minute, ':', ...second] : [...hour, ':', ...minute]
  ), [displaySeconds, hour, minute, second]);
```

File: components/bl-worldClock-component/src/views/digital-types/DigitalClockPixel.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  return (
+    <div className="clock">
+      { digits.map((value, index) => (
+        <Digit key={ index } value={ value } colors={ colors } />
```
Comment (serhmel) [inline components/bl-worldClock-component/src/views/digital-types/DigitalClockPixel.js:46]:
```suggestion
        <Digit key={ index } value={ value } colors={ colors }/>
```

File: components/bl-worldClock-component/styles/index.less
Comment (serhmel) [inline components/bl-worldClock-component/styles/index.less:1]: font-weight property is duplicated

```
      .ampm {
        align-self: flex-end;
        color: @bl-customComponent-worldClock-digital-pixel-ampm-color;
        font-weight: @bl-customComponent-worldClock-digital-pixel-ampm-font-weight;
        line-height: @bl-customComponent-worldClock-digital-pixel-ampm-line-height;
        font-size: @bl-customComponent-worldClock-digital-pixel-ampm-font-size;
        font-weight: @bl-customComponent-worldClock-digital-pixel-ampm-font-weight;
        margin-left: @bl-customComponent-worldClock-digital-pixel-ampm-margin;
      }
```

File: components/bl-worldClock-component/src/views/analog-clock.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+export function AnalogClock({ time, label, type, displaySeconds }) {
+  const { hour, minute, second } = time;
+
+  const Clock = AnalogClockViews[type];
+
+  return (
+    <div className={ cn('analog', type) }>
+      { Clock ? (
+        <Clock handStyle={ useHandStyles(hour, minute, second, displaySeconds) }/>
+      ) : (
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/analog-clock.js:25]: Component could exists without type here? How does it happens?

File: components/bl-worldClock-component/src/views/digital-clock.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+export function DigitalClock({ time, label, type, displaySeconds }) {
+  const Clock = DigitalClockViews[type];
+
+  return (
+    <div className={ cn('digital', type) }>
+      { Clock && (
+        <Clock time={ time } displaySeconds={ displaySeconds }/>
+      ) }
+
+      <div className="clock-label"><span className="clock-label-text">{ label }</span></div>
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/digital-clock.js:30]: <Label> { content } <Label/>

File: components/bl-worldClock-component/src/views/digital-types/DigitalClockFlip.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+function FlipUnitContainer({ digit, shuffle, unit }) {
+  let currentDigit, previousDigit;
+
+  if (unit === Formats.AMPM) {
+    currentDigit = digit;
+    previousDigit = digit === Formats.AM ? Formats.PM : Formats.AM;
+  } else {
+    currentDigit = digit * 1;
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/digital-types/DigitalClockFlip.js:85]: Is it works like a `Number('11') => 11`? 

If yes ,change it please. In other case explain how it works.

File: components/bl-worldClock-component/src/views/digital-types/DigitalClockPixel.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  '2': [1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1],
+  '3': [1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1],
+  '4': [1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1],
+  '5': [1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1],
+  '6': [1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1],
+  '7': [1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1],
+  '8': [1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1],
+  '9': [1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1],
+  ':': [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
+};
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/digital-types/DigitalClockPixel.js:15]: We still works with it? Or we waiting for removing it?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-worldClock-component/src/views/analog-clock.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+  return { hour, minute, ...(displaySeconds && { second }) };
+}
+
+function useDate(hour, minute, second) {
+  const date = new Date();
+
+  date.setHours(hour, minute, second);
+
+  return date;
+}
```
[the remark this answers is not in the document]
Comment (Fialkaksu) [reply components/bl-worldClock-component/src/views/analog-clock.js:107]: Unfortunately, we cannot use the valueOf of the dayjs date object, it returns the local time anyway.

File: components/bl-worldClock-component/src/views/analog-clock.js (hunk 7 above)
Comment (Fialkaksu) [reply components/bl-worldClock-component/src/views/analog-clock.js:25]: The type for the component is defined above before the return

File: components/bl-worldClock-component/src/views/digital-clock.js (hunk 8 above)
Comment (Fialkaksu) [reply components/bl-worldClock-component/src/views/digital-clock.js:30]: Sorry, not content, but a label, a label that is displayed under the clock.

File: components/bl-worldClock-component/src/views/digital-types/DigitalClockFlip.js (hunk 9 above)
Comment (Fialkaksu) [reply components/bl-worldClock-component/src/views/digital-types/DigitalClockFlip.js:85]: Yes, it works like a `Number ('11') => 11`. Fixing it

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-worldClock-component/README.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+Below is a Codeless Example highlighting how to use the World Clock component:
+
+<p align="center">
+  <img src="./example-images/settings.png" alt="settings" width="500"/>
+</p>
+
+<p align="center">
+  <img src="./example-images/component-preview.png" alt="component preview" width="500"/>
+</p>
```
Comment (v-excelsior) [inline components/bl-worldClock-component/README.md:167]: Unfortunately, this guide doesn't show anything that could help.

I'm not talking about removing it, but if you wan't to leave that - make properties image smaller, and add subheading for example

```
## Examples

### Usage
Below is a Codeless Example highlighting how to use the World Clock component:
...

```

File: components/bl-worldClock-component/component.json [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+      "name": "label",
+      "label": "Label",
+      "showInSettings": true,
+      "hasLogicHandler": true,
+      "handlerId": "labelLogic",
+      "defaultValue": "London",
+      "handlerLabel": "Label Logic",
+      "dataBinding": true,
+      "settingTooltip": "The name of a location/city/etc to show in the component.",
+      "handlerDescription": "This is a handler for the logic to determine the name of a location/city/etc to display it in the component."
```
Comment (v-excelsior) [inline components/bl-worldClock-component/component.json:178]: This is a handler for the logic to determine the content of Clock label.

File: components/bl-worldClock-component/component.json [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+    },
+    {
+      "type": "checkbox",
+      "name": "displaySeconds",
+      "label": "Display Seconds",
+      "showInSettings": true,
+      "defaultValue": true,
+      "hasLogicHandler": false,
+      "handlerId": "displaySecondsLogic",
+      "settingTooltip": "Select to display the seconds(second hand)."
```
Comment (v-excelsior) [inline components/bl-worldClock-component/component.json:208]: `Show the second hand when enabled.`

or

`Disable to hide second hand.`

Or any other variation without "Select"

File: components/bl-worldClock-component/component.json [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    }
+  ],
+  "eventHandlers": [],
```
Comment (v-excelsior) [inline components/bl-worldClock-component/component.json:211]: As idea we could add events that runs every second/minute/hour. What you think about it?

File: components/bl-worldClock-component/src/views/analog-clock.js [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+  const { hour, minute, second } = time;
+
+  const clockProps = useMemo(() => {
+    if (clockStyle === AnalogClockTypes.CLASSIC) {
+      return {
+        date           : useDate(hour, minute, second),
+        secondHandColor: displaySeconds ? 'red' : 'transparent',
+      };
+    } else {
+      return {handStyle: useHandStyles(hour, minute, second, displaySeconds)};
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/analog-clock.js:31]: formatting

File: components/bl-worldClock-component/src/index.js [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+  const [clockType, clockStyle] = type.split('-');
+  const WorldClock = ClockViews[clockType];
+
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <div className={ cn('bl-customComponent-worldClock', classList) } style={ style } ref={ elRef }>
+      { WorldClock && (
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/index.js:50]: Same question. Why we need this check?

File: components/bl-worldClock-component/src/views/analog-clock.js [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+      <div className="clock-rounder"/>
+    </div>
+  );
+}
+
+function Markers({ markersFull=false }) {
+  return (
+    <>
+      { markersFull ? (
+        [...Array(12).keys()].map((i, key) => {
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/analog-clock.js:86]: `key` and `i` is same, isn't it? We really need `key`?

File: components/bl-worldClock-component/src/views/analog-clock.js [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+      })) : (
+        <>
+          <span className="clock-twelve"/>
+          <span className="clock-three"/>
+          <span className="clock-six"/>
+          <span className="clock-nine"/>
+        </>
+      ) }
+    </>
+  );
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/analog-clock.js:104]:
```suggestion
if(markersFull) {
  return <..>
}

return <...>
```

Try to use as less as possible logic in JSX.

File: components/bl-worldClock-component/src/views/analog-clock.js [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+  date.setHours(hour, minute, second);
+
+  return date;
+}
+
+function useHandStyles(hours, minutes, seconds, displaySeconds) {
+  const hourDegree = (hours % 12) * 30 + minutes * 0.5;
+  const minuteDegree = minutes * 6 + seconds * 0.1;
+  const secondDegree = seconds * 6;
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/analog-clock.js:118]: empty line

File: components/bl-worldClock-component/src/views/digital-clock.js [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+};
+
+export function DigitalClock({ time, label, clockStyle, displaySeconds }) {
+  const Clock = DigitalClockViews[clockStyle];
+
+  return (
+    <div className={ cn('digital', clockStyle) }>
+      { Clock && (
+        <Clock time={ time } displaySeconds={ displaySeconds }/>
+      ) }
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/digital-clock.js:29]: Why we need this check?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-worldClock-component/src/views/digital-types/digital-clock-flip.js [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+export function DigitalClockFlip({ time, displaySeconds }) {
+  const { hour, minute, second, isAmpm, ampm } = time;
+
+  const [state, setState] = useObjectState({
+    hour, minute, second, ampm,
+    hourShuffle  : true,
+    minuteShuffle: true,
+    secondShuffle: true,
+    ampmShuffle  : true,
+  });
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/digital-types/digital-clock-flip.js:30]: Could we write like this?

```suggestion
  const [ {hour, minute, second, ampm, hourShuffle, minuteShuffle, secondShuffle, ampmShuffle },setState]=seObjectState({
    ...time,
    hourShuffle  : true,
    minuteShuffle: true,
    secondShuffle: true,
    ampmShuffle  : true,
  });
```

And below and in components change

`hour` -> `time.hour` and `state.hour` - `hour`

File: components/bl-worldClock-component/src/views/digital-types/digital-clock-flip.js [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+  let currentDigit, previousDigit;
+
+  if (unit === Formats.AMPM) {
+    currentDigit = digit;
+    previousDigit = digit === Formats.AM ? Formats.PM : Formats.AM;
+  } else {
+    currentDigit = Number(digit);
+    previousDigit = digit - 1;
+
+    [previousDigit, currentDigit] = ensureLimits(previousDigit, currentDigit, unit);
```
Comment (v-excelsior) [inline components/bl-worldClock-component/src/views/digital-types/digital-clock-flip.js:88]: I think it's too should work, as we have mutation inside the ensureLimits function

```suggestion
    ensureLimits(previousDigit, currentDigit, unit);
```

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-worldClock-component/src/views/digital-types/digital-clock-flip.js (hunk 23 above)
Comment (Fialkaksu) [reply components/bl-worldClock-component/src/views/digital-types/digital-clock-flip.js:88]: Sorry, that doesn't work.

**Pull request merged**